File: Year-2023/LeetCode-March-Challenge/LinkedListCycle2.cpp
#include<bits/stdc++.h>
using namespace std;
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *slow = head, *fast = head;
        while(fast && fast->next){
            slow = slow->next;
            fast = fast->next->next;
            if(slow == fast) break;
        }

        if(!fast || !fast->next) return nullptr;
        fast = head;
        while(fast != slow){
            slow = slow->next;
            fast = fast->next;
        }
        return slow;
    }
};

File: Year-2023/LeetCode-March-Challenge/LinkedListCycle2.js
function ListNode(val) {
    this.val = val;
    this.next = null;
}

var detectCycle = function (head) {
    let slow = head, fast = head;

    while (fast && fast.next) {
        slow = slow.next;
        fast = fast.next.next;
        if (slow === fast) break;
    }

    if (!fast || !fast.next) return null;
    fast = head;

    while (fast != slow) {
        slow = slow.next;
        fast = fast.next;
    }
    return slow;
};